File: main/delete-node-in-a-bst/delete-node-in-a-bst-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode DeleteNode(TreeNode root, int key) {
        if (root == null) return null;

        if (key < root.val) {
            root.left = DeleteNode(root.left, key);
            return root;
        }

        if (root.val < key) {
            root.right = DeleteNode(root.right, key);
            return root;
        }

        // If it has at most than one child just replace it with that (or null).
        if (root.right == null) return root.left;
        if (root.left == null) return root.right;

        // If its right child has no left child, attach its left child as the
        // left chld of its right child.
        if (root.right.left == null) {
            root.right.left = root.left;
            return root.right;
        }

        // Find the minimal node in its right subtree.
        var parent = root.right;
        var child = parent.left;
        while (child.left != null) {
            parent = child;
            child = child.left;
        }

        // Replace it with that minimal node. Very carefully.
        parent.left = child.right;
        child.right = root.right;
        child.left = root.left;
        return child;
    }
}

File: main/delete-node-in-a-bst/delete-node-in-a-bst-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode deleteNode(final TreeNode root, final int key) {
        final TreeNode meta = new TreeNode(Integer.MIN_VALUE);
        meta.right = root;

        final TreeLinkHandle pos = new TreeLinkHandle(meta, Side.RIGHT);
        descendTo(pos, key);

        if (pos.get() != null) {
            if (pos.get().left == null)
                pos.set(pos.get().right);
            else if (pos.get().right == null)
                pos.set(pos.get().left);
            else
                replace(pos, extractNextBelow(pos));
        }

        return meta.right;
    }

    private static void descendTo(final TreeLinkHandle pos, final int key) {
        while (pos.get() != null && pos.get().val != key)
            pos.go(key < pos.get().val ? Side.LEFT : Side.RIGHT);
    }

    private static void descendToMin(final TreeLinkHandle pos) {
        while (pos.get().left != null) pos.go(Side.LEFT);
    }

    private static TreeNode extractNextBelow(TreeLinkHandle pos) {
        pos = pos.copy();
        pos.go(Side.RIGHT);
        descendToMin(pos);

        final TreeNode ret = pos.get();
        pos.set(ret.right);
        return ret;
    }

    private static void replace(final TreeLinkHandle pos, final TreeNode node) {
        node.left = pos.get().left;
        node.right = pos.get().right;
        pos.set(node);
    }
}

/** Identifies which direction to traverse in a binary tree. */
enum Side { LEFT, RIGHT }

/** A re-seatable handle to use or change a child pointer in a binary tree. */
final class TreeLinkHandle {
    TreeLinkHandle(final TreeNode parent, final Side side) {
        _parent = parent;
        _side = side;
    }

    TreeNode get() {
        switch (_side) {
        case LEFT:
            return _parent.left;
        case RIGHT:
            return _parent.right;
        }

        throw new AssertionError("Internal error: can't read invalid side!");
    }

    void set(final TreeNode node) {
        switch (_side) {
        case LEFT:
            _parent.left = node;
            return;
        case RIGHT:
            _parent.right = node;
            return;
        }

        throw new AssertionError("Internal error: can't write invalid side!");
    }

    void go(final Side side) {
        _parent = get();
        _side = side;
    }

    TreeLinkHandle copy() {
        return new TreeLinkHandle(_parent, _side);
    }

    private TreeNode _parent;

    private Side _side;
}
